components/renku_data_services/metrics/orm.py

```diff
@@ -0,0 +1,40 @@
+"""SQLAlchemy schemas for the metrics database."""
+
+from datetime import datetime
+from typing import Any, Optional
+
+from sqlalchemy import JSON, DateTime, Identity, Integer, MetaData, String, func
+from sqlalchemy.dialects.postgresql import JSONB
+from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass, mapped_column
+
+JSONVariant = JSON().with_variant(JSONB(), "postgresql")
+
+
+class BaseORM(MappedAsDataclass, DeclarativeBase):
+    """Base class for all ORM classes."""
+
+    metadata = MetaData(schema="metrics")
+
+
+class MetricsORM(BaseORM):
+    """Metrics staging table.
+
+    Events are stored in this table and then processed by a background task that sends them to the actual metrics
+    service.
+    """
+
+    __tablename__ = "metrics"
+
+    id: Mapped[int] = mapped_column(Integer, Identity(always=True), primary_key=True, default=None, init=False)
```

Comment:
question: Why not ulid? Since we will read and write a lot. I am not sure how postgres (if at all) reclaims the range for the index. And not sure if we can hit the max value that `Integer` can hold in postgres. With ulid we dont have to worry about this. We added a `generate_ulid()` function in postgres. So you can get postgres to generate a ulid for you by having `server_default="generate_ulid()"`.

See https://github.com/SwissDataScienceCenter/renku/blob/master/scripts/init-db/generate_ulid_func.sql

That function is already added in the db. And also in the devcontainer too.

Reply:
Assuming that we send 1,000,000 events per month (the free tier limit) it takes couple of thousands months to reach that limit and integers have less overhead compared to ULID. I absolutely agree with you though, using ULID won't cause possible headaches in future and I've done what you suggested.